core/src/main/scala/org/apache/spark/storage/BlockManagerDecommissioner.scala

```diff
@@ -327,4 +354,28 @@ private[storage] class BlockManagerDecommissioner(
     }
     logInfo("Stopped storage decommissioner")
   }
+
+  /*
+   *  Returns the last migration time and a boolean for if all blocks have been migrated.
+   *  If there are any tasks running since that time the boolean may be incorrect.
+   */
+  private[storage] def lastMigrationInfo(): (Long, Boolean) = {
+    if (stopped || (stoppedRDD && stoppedShuffle)) {
+      (System.nanoTime(), true)
+    } else {
+      // Chose the min of the running times.
+      val lastMigrationTime = if (
+        conf.get(config.STORAGE_DECOMMISSION_SHUFFLE_BLOCKS_ENABLED) &&
+        conf.get(config.STORAGE_DECOMMISSION_RDD_BLOCKS_ENABLED)) {
+        Math.min(lastRDDMigrationTime, lastShuffleMigrationTime)
```

Comment:
This must be `Math.max`.

As I understand in `CoarseGrainedExecutorBackend` the `lastTaskRunningTime` increases with about 1000ms in every iteration, let's assume exactly 1000ms. So if the RDD migration finished at 500ms (let's count in the example from 0 here) but shuffle files to be migrated are still left and they will be finished only in the next round (let's assume in 1500ms) then we we never shutdown the executor: as in the current round `blocksMigrated` is false and in all the following ones `migrationTime` will be less than `lastTaskRunningTime`, so this condition will be never satisfied:

https://github.com/apache/spark/blob/484f8e216d5727e516488aedbdb41b1f63569701/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L305

Reply:
I was thinking a bit more on this part. 

The intention of `lastRDDMigrationTime`, `lastShuffleMigrationTime` and `lastTaskRunningTime` is to notify the driver **only once** about the exit of this executor. 

My first thought was to improve this by expressing this intention directly with a simple flag in the `shutdownThread`(like `exitTriggered` which can be used as a stopping condition instead of the `while (true)` ) and remove all the three time variables and change `lastMigrationInfo` to return just a boolean and rename it for its new role (for example to `isFinished` or just simply to `finished`). 

Then I went a bit further and checked the `exitExecutor` method and I think even the flag is not necessarily needed:
https://github.com/apache/spark/blob/484f8e216d5727e516488aedbdb41b1f63569701/core/src/main/scala/org/apache/spark/executor/CoarseGrainedExecutorBackend.scala#L267

Still I like to have a stopping condition in the while loop so please consider to use the flag instead of the times.

Reply:
@holdenk @agrawaldevesh what do you think?

Reply:
I agree with you @attilapiros. On both the counts:
- Converting this into a level trigger vs an edge trigger: ie the flag goes on and remains on when it is time to exit.
- Having consistency b/w the termination condition of the thread above and the computation of the boolean variable in this method. Having a helper method would help with this and also improve longer term maintainability.

IMHO, this is great feedback and will improve this PR. Thank you.

Reply:
So it only increases if there is a task running. We don’t want it to shutdown or consider the blocks migrated if there is a task migrated.
Unfortunately this flag can come and go depending on the tasks running on the executor which is why it’s structured the way it is.

Reply:
Now I am starting to get this part. Can we try to simplify this? :)

On a decommissioned executor there will be no new tasks started, right? (Theoretically one task could be started as the scheduler does not processed the `DecommissionExecutor` message but let's take this corner case out as we can have this first 1 sec sleep which more or less handles this).

So when `executor.numRunningTasks` will be 0 it stays 0. As I remember caching is part of task running so no new cached blocks will be created when numRunningTasks=0.

So would it work if we first wait to reach numRunningTasks==0 in a sleep check loop and then we would check for the migration finished flag without using the time part of `lastMigrationInfo`?

Reply:
So we don’t reject tasks sent to us and an executor can start decommissioning without the driver knowing yet so it is possible (although unlikely) to get a new task after we’ve started decommissioning.

Reply:
Yes, I know this. That is what I meant by

>  (Theoretically one task could be started as the scheduler does not processed the DecommissionExecutor message but let's take this corner case out as we can have this handled with some sleep).

So I think with some waiting (which I would avoid if I could) the probability of this unlikely case can be decreased more.

But let's assume even that very unlikely case happens then we are losing one cached RDD from the current stage which will be re-submitted. The advantage on the other hand a more simple code which can be fixed/improved by more developers.

**Other idea**: leave this logic as it is but refactor the `shutdownThread` into a separate class to make it unittestable (all its dependencies would be mockable, even the time its uses would be Clock a ManualClock in test).

What do you think?

Reply:
Oh sorry I missed that part. So currently the block manager doesn't know if there are any running tasks, and I'm not sure it's better to expose that into the block manager.

I'd rather not unit test the shutdownThread, realistically I think it's better suited to the integration style testing that we do. I don't mind moving it into a separate class/file though.

It seems like your advocating for either a solution with more edge cases or less information hiding which is why I'm hesitation to pursue either of these paths. The time stamp is certainly a bit complicated, but I'm not convinced it's complicated enough to throw it out and throw away good data for the user. What about if we documented how the timestamps are used a bit more would that be enough?

Reply:
Just a little misunderstanding remained:
> So currently the block manager doesn't know if there are any running tasks, and I'm not sure it's better to expose that into the block manager.

I do not think that is needed: I would implement the simplified solution within the shutdown thread.

I advise some improved robustness by a more testable or easier provable code (of which easier to argue about) to make it more suppartable/extendable. 

I am sorry but regarding this code part your reviewer(s) (at least me myself) failed at first :( that is an indication we should do something. Yes we can try to improve the documentation first. 

@agrawaldevesh: what is your opinion/idea?

Reply:
Ok I'm not sure how we'd implement this in the shutdown thread, maybe that's why I'm concerned about the proposal. In your mind, information from the block manager decommissioner would we expose to allow this logic to move to the shutdown thread?

Reply:
Also by "failed at first" what are you referring to? That the decommission call would block on the RPC thread or some other bug?

Reply:
On Monday I can open a PR what I had in my mind (that is my last day before a two-three weeks long computerless period). 

Regarding failed at first: I am referring to my mistake. To the fact I misunderstand your intentions regarding the code. So I was completely wrong here:
https://github.com/apache/spark/pull/29211#discussion_r460263043

Reply:
PR against your branch

Reply:
> @agrawaldevesh: what is your opinion/idea?

I confess that even I am finding the logic in the lastMigrationInfo a bit hard to follow. And as far as I can see, this logic is the key to getting the shutdown thread to exit cleanly. If the information returned by lastMigrationInfo is wrong then the shutdown thread may exit prematurely or never. Both of which should be avoided, particularly because the shutdown thread has no timeout on the amount of time it can hang around. 

This complexity is intrinsic. The logic is indeed complex but I think what we can add some more documentation here, perhaps explaining the intent of the code using a couple of examples ? Even better would be to please add a unit test for this. I think it should be easy to unit test just the lastMigrationInfo. This is just testing that the function is behaving as expected by stubbing out other parts. This test would also help with providing the necessary guardrails as we change this function in the future. I expect this function to change as we production harden the cache/shuffle block migration in the near term.

Reply:
Yeah I’m down to unit test this function, makes sense to me and it’s fairly standalone.

I’ll work on the documentation over the weekend.

Reply:
https://github.com/holdenk/spark/pull/7 :)

Reply:
So the approach in that PR re-opens the race condition were preventing here. I would really rather not do that.

I’d also like us to make progress here though, so we can temporarily accept the race condition and file a JIRA and revisit it later as a stand-alone item if y’all are not comfortable with any of the ways to avoid the race.